lib/extract-stream.js

```diff
@@ -15,11 +15,10 @@ function extractStream (dest, opts) {
     uid: opts.uid,
     gid: opts.gid,
     onentry (entry) {
-      if (entry.type.toLowerCase() === 'file') {
-        entry.mode = opts.fmode & ~(opts.umask || 0)
-      } else if (entry.type.toLowerCase() === 'directory') {
-        entry.mode = opts.dmode & ~(opts.umask || 0)
-      }
+      entry.mode = entry.mode | (
+        entry.type.toLowerCase() === 'directory' ? opts.dmode : opts.fmode
+      )
+      entry.mode = entry.mode & (~opts.umask || 0)
```

Comment:
Came here to say that `~(opts.umask || 0)` isn't the same as `(~opts.umask || 0)`, but for all intents and purposes, it is 0_o